Feature structuredformatstring

Praburaj opened the pull request:
This adds ability to specify named placeholders while writing logs. 

For example: 
`logger.WriteInformation("Tests started at '{StartTime}'", DateTime.Now)` (instead of `'{0}'` => `'{StartTime}'`). 

@lodejard

Praburaj commented on src/Microsoft.Framework.Logging/LoggerStructureFormat.cs:
```diff
@@ -0,0 +1,108 @@
+                    var openBraceIndex = FindIndexOf(format, '{', scanIndex, endIndex);
+                    var closeBraceIndex = FindIndexOf(format, '}', openBraceIndex, endIndex);
+
+                    // Format item syntax : { index[,alignment][ :formatString] }.
```
@lodejard I had to update this as the format item syntax is per the comment here.

Eilon commented on src/Microsoft.Framework.Logging/LoggerStructureFormat.cs:
```diff
@@ -0,0 +1,105 @@
+
+        public LoggerStructureFormat(string format, params object[] values)
+        {
+            if (!_formatters.TryGetValue(format, out _formatter))
```
Why not just always call `GetOrAdd` with a `Func`?

Eilon replied:
This method: https://msdn.microsoft.com/en-us/library/ee378677(v=vs.110).aspx

Praburaj replied:
Good point.

Eilon commented on src/Microsoft.Framework.Logging/LoggerStructureFormat.cs:
```diff
@@ -0,0 +1,105 @@
+            {
+                var sb = new StringBuilder();
+                var endIndex = format.Length;
+                for (var scanIndex = 0; scanIndex != endIndex;)
```
Looks like a `while` loop to me.

Praburaj replied:
yeah..

Eilon commented on src/Microsoft.Framework.Logging/LoggerStructureFormat.cs:
```diff
@@ -0,0 +1,105 @@
+            return new Formatter(format);
+        }
+
+        class Formatter
```
Fix coding style throughout.

Praburaj replied:
I decorated all methods with `public/private/internal` - are there other specific comments?

Eilon commented on src/Microsoft.Framework.Logging/LoggerStructureFormat.cs:
```diff
@@ -0,0 +1,105 @@
+            return _formatter.GetValues(_values);
+        }
+
+        private Formatter ParseFormatter(string format)
```
Is this method needed?

Praburaj replied:
Can be removed. I updated.

Eilon commented on samples/SampleApp/Program.cs:
```diff
@@ -19,15 +28,27 @@ public Program()
 #endif
-            factory.AddConsole();
-            factory.AddConsole((category, logLevel) => logLevel >= LogLevel.Critical && category.Equals(typeof(Program).FullName));
+            //factory.AddConsole();
```
Why are these commented? Perhaps I'm confused about the scenario for this but I thought it was to enable structure log formatting to the built-in loggers, such as the Console logger?

Praburaj replied:
This shouldn't really be commented. I'll uncomment this. I just took over code and missed to notice this.

Praburaj replied:
This feature adds ability to specify named placeholders as I mentioned in the PR description.

Eilon commented on src/Microsoft.Framework.Logging/LoggerStructureFormat.cs:
```diff
@@ -0,0 +1,99 @@
+
+        public LoggerStructureFormat(string format, object[] values)
+        {
+            _formatter = _formatters.GetOrAdd(format, _ => new Formatter(format));
```
Curious - can you pass `format` in as the parameter to the lambda? Otherwise it looks like it's putting `format` in the closure, which I don't think is intentional (or needed) here. Not 100% sure about this though.

Praburaj replied:
We can't pass `format` as a parameter to the lambda.

Eilon commented on src/Microsoft.Framework.Logging/LoggerStructureFormat.cs:
```diff
@@ -0,0 +1,99 @@
+                _format = sb.ToString();
+            }
+
+            private int FindIndexOf(string format, char ch, int startIndex, int endIndex)
```
Can make `static`?

Praburaj replied:
yeah. sure.

Eilon commented on src/Microsoft.Framework.Logging/LoggerStructureFormat.cs:
```diff
@@ -0,0 +1,99 @@
+
+namespace Microsoft.Framework.Logging
+{
+    public class LoggerStructureFormat : ILoggerStructure
```
Maybe add a doc comment to this?

Praburaj commented:
Updated PR.

avanderhoorn commented:
Really like this PR!!!! Its been on our wish list for ever but never thought it would make it into the core platform. 

We are wanting to do things where we visually tokenise the trace messages we are showing and describe to people what they are seeing. With the options we have in `LoggerStructureFormat` currently, we probably can't do this however, as we only have access to the finial "formatted" or the list of key/value tokens. This isn't enough on its own for us to identify the tokens. Any chance we can have access to the underlying original format string as well? 

This would mean that when we visually show things, we can locate the correct places for highlighting.

PureKrome commented:
![](https://31.media.tumblr.com/751b9113f1b9084c2858d9ca8129c260/tumblr_inline_n4rt6dgDo51solpjm.gif)

Praburaj commented:
@avanderhoorn is the ask to just expose the original format string as a property on `LoggerStructureFormat`?

avanderhoorn commented:
Yes. I think thats all I would need to identify the tokens and build what I need. 

If i was making wishes to santa however, instead of me need to do that manually, it would be amazing if we could get the positions of the tokens... as at the moment, if we want to show the visualized version of the tokenized string, we have to recreate the logic in `LoggerStructureFormat` in order to put the `span`s around each token, etc. Does that make sense?

Eilon commented on src/Microsoft.Framework.Logging/LoggerStructureFormat.cs:
```diff
@@ -0,0 +1,103 @@
+
+                while (scanIndex < endIndex)
+                {
+                    var openBraceIndex = FindIndexOf(format, '{', scanIndex, endIndex);
```
I think we need to consider how to escape curly braces. Currently I think it's broken if you try to put an arbitrary curly brace in the format. The typical way to escape curly braces is to double them, i.e. `{{` and `}}`. That's pretty much what `string.Format` does and also ASP.NET Routing.

lodejard replied:
Good catch!

Praburaj replied:
I will do a bit more testing on this and update the PR.

lodejard commented:
@avanderhoorn I like the idea of having the format string as a property, though not all messages which have structure will have a format string of course. How about if the code that parses and renders the message was public and exposed the info you described?

avanderhoorn commented:
Makes sense on the former (and I agree) and the latter would be great!!!
I'm trying to get to be point were it is as simple as possible to work out
what portions of the string where tokenized and where those tokens are (so
I can do the text highlighting) - showing that info would be a great step
in that direction.

On Saturday, January 31, 2015, Louis DeJardin <notifications@github.com>
wrote:

> @avanderhoorn <https://github.com/avanderhoorn> I like the idea of having
> the format string as a property, though not all messages which have
> structure will have a format string of course. How about if the code that
> parses and renders the message was public and exposed the info you
> described?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/aspnet/Logging/pull/98#issuecomment-72330017>.
>

Praburaj commented:
@Eilon @loudej @avanderhoorn - I have updated the PR addressing all the issues and adding more tests to cover the brace escape cases. 

@avanderhoorn I'm exposing the original format string as an item from the values returned from `GetValues()` with a special key named `__OriginalFormat`. I'm open to feedback on this.

avanderhoorn commented:
Thanks for getting this data out! In terms of feedback, I must be honest, accessing it via that key seems a little unintuitive and more like a private api... Would be nice if there was some more supported way of doing this.

Eilon commented on test/Microsoft.Framework.Logging.Test/LoggerStructureFormatTest.cs:
```diff
@@ -0,0 +1,61 @@
+
+namespace Microsoft.Framework.Logging.Test
+{
+    public class LoggerStructureFormatTest
```
Can you add tests for un-balanced cases? E.g. various bad strings such as `{foo` and `bar}`.

Praburaj replied:
There are already some tests on these unbalanced cases but escaped braces. Added some more negative ones.

Praburaj commented:
@loudej @avanderhoorn 

>Thanks for getting this data out! In terms of feedback, I must be honest, accessing it via that key seems a little unintuitive and more like a private api... Would be nice if there was some more supported way of doing this.

How about adding a property as well on the Formatter to give the original format string apart from this?

avanderhoorn commented:
I like the addition of the first class property. I'm still left scratching my head though over why we would need the `__OriginalFormat` in the `GetValues()`. As a caller of `GetValues()` I kinda expect that the collection I get back to be a collection of `keys` and `values` that I found in the `OriginalFormat`. I wouldn't expect one of the entries to be something that wasn't a found `key`/`value`... I'm not saying that its not useful, just that I doesn't really make sense given what the "contract" describes.

Eilon commented:
:+1: to what @avanderhoorn said. The original format string is not a real value, so we shouldn't put a fake value in the list.

Praburaj commented:
@loudej suggested to do this as if someone has a reference of `ILoggerStructure` one can directly get the original format by calling `GetValues()`. If its just a property one has to typecast the reference to the specific type to get the original format string. Exposing a property as well doesn't hurt if you have an instance of the `LoggerStructureFormatter` itself. So I ended up adding both.

yishaigalatzer commented:
@Praburaj you can probably add `{OriginalFormat}` as the key instead of `__OriginalFormat` ;)

Praburaj commented:
@yishaigalatzer yeah that's doable.

avanderhoorn commented:
What if `OriginalFormat` is a key the user already has?

Praburaj commented:
yeah. we'll have to special case this key. User cannot have this key name.

lodejard commented:
I think Yishai meant the key name would literally contain curly braces, which end-user format strings could not reproduce.

yishaigalatzer commented:
Yes exactly that

-----Original Message-----
From: "Louis DeJardin" <notifications@github.com>
Sent: ‎2/‎3/‎2015 7:17 PM
To: "aspnet/Logging" <Logging@noreply.github.com>
Cc: "Yishai Galatzer" <yishai_galatzer@yahoo.com>
Subject: Re: [Logging] Feature structuredformatstring (#98)

I think Yishai meant the key name would literally contain curly braces, which end-user format strings could not reproduce.
—
Reply to this email directly or view it on GitHub.

Praburaj commented:
Addressed all the comments. If there are no other comments this PR can go in.

Eilon commented:
:shipit: If others have more comments, please log bugs so we can track those ideas separately. Thanks!

Praburaj commented:
Thanks everyone.
